### ISSUE TITLE
Questions about the 'write_copies' function!

### ISSUE THREAD

**Issue by geekgeekhello** [color/clrsvsim#5]
In the 'write_copies' function, the code for downsampling reads is as below:
        # Probabilistically write the remaining fractional part of num_copies.
        if (num_copies == .5 and hash(read.qname) % 2) or random.random() < num_copies:
            outsam.write(read)
            copies_written += 1

Why does it have this condition check '(num_copies == .5 and hash(read.qname) % 2)' specifically for ratio_change=0.5? This would result in ~75% of the original reads be written to the simulated BAM when ratio_change=0.5. Is it intentional? What is the biology behind this?

Thanks

**Comment by giladmishne** [comment]
@alexrson can you explain commit https://github.com/color/clrsvsim/pull/3/commits/df334a34b5ecd96a1125a3d570184c8ef78e7d16#diff-cead38915fa9995344348323366bce56R253 ?

**Comment by alexrson** [comment]
Yeah, I wanted the two mates to be either both present or both absent otherwise converting to fastq doesn't work well. So for the fractional portion I changed from using the random function to using the mod of the hash, essentially random function that will always give the same result for both mates. It could be better to just specify seed maybe.

Is that the change you are asking about?

**Comment by geekgeekhello** [comment]
@alexrson Thanks for the explanation. I still have a couple of questions.
My understanding is that, with that line of code (see pasted below, it is Line 253 in simulator.py):
--------------------------------------------------------
if (num_copies == .5 and hash(read.qname) % 2) or random.random() < num_copies:
--------------------------------------------------------
(1) The 'mod of hash' is applied only when num_copies (ratio_change)=0.5. e.g., if num_copies=0.75 or 0.6, only the random function is applied. Why is that?
(2) When num_copies = 0.5, both mod of hash and the random function are applied, which will result in roughly 75% of reads written to the simulated BAM, which is significantly higher than the expected ratio. Is that intentional?

Thanks!

**Comment by alexrson** [comment]
I think you're correct that that is a bug. I'll submit a fix as soon as I get a chance.
